File: packages/python/helix_sdk/transport/sse.py
# packages/python/helix_sdk/transport/sse.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, AsyncIterator

import httpx

from ..constants import HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION, run_events_path
from ..errors import AEPTransportError


@dataclass
class SseEvent:
    event_id: str
    run_id: str
    type: str
    emitted_at: str
    payload: Any


class SseClient:
    def __init__(self, base_url: str, bearer: str, http_client: httpx.AsyncClient | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._bearer = bearer
        self._client = http_client or httpx.AsyncClient(timeout=httpx.Timeout(60.0, read=None))
        self._owns_client = http_client is None

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def stream(
        self,
        run_id: str,
        session_id: str | None,
        since: str | None = None,
    ) -> AsyncIterator[SseEvent]:
        headers = {
            "Accept": "text/event-stream",
            "Authorization": f"Bearer {self._bearer}",
            HEADER_PROTOCOL_VERSION: PROTOCOL_VERSION,
        }
        if session_id:
            headers[HEADER_SESSION_ID] = session_id

        params = {"since": since} if since else None

        async with self._client.stream(
            "GET",
            f"{self._base_url}{run_events_path(run_id)}",
            headers=headers,
            params=params,
        ) as resp:
            if resp.status_code != 200:
                raise AEPTransportError(f"SSE HTTP {resp.status_code} {resp.reason_phrase}")

            buf = ""
            async for chunk in resp.aiter_text():
                buf += chunk
                while "\n\n" in buf:
                    frame, buf = buf.split("\n\n", 1)
                    parsed = _parse_frame(frame)
                    if parsed is not None:
                        yield parsed


def _parse_frame(raw: str) -> SseEvent | None:
    event_type = "message"
    data_line = ""
    for line in raw.split("\n"):
        if line.startswith("event:"):
            event_type = line[6:].strip()
        elif line.startswith("data:"):
            data_line += line[5:].strip()
    if not data_line:
        return None
    try:
        parsed = json.loads(data_line)
    except json.JSONDecodeError:
        return None
    return SseEvent(
        event_id=str(parsed.get("event_id", "")),
        run_id=str(parsed.get("run_id", "")),
        type=str(parsed.get("type", event_type)),
        emitted_at=str(parsed.get("emitted_at", "")),
        payload=parsed.get("payload", parsed),
    )

File: packages/typescript/src/transport/sse.ts
// packages/typescript/src/transport/sse.ts
import { AEPTransportError } from "../errors.js";
import { HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION, RUN_EVENTS_PATH } from "../constants.js";

export interface SseEvent {
  event_id: string;
  run_id: string;
  type: string;
  emitted_at: string;
  payload: unknown;
}

export interface SseOptions {
  since?: string;
  signal?: AbortSignal;
}

export class SseClient {
  private readonly baseUrl: string;
  private readonly bearer: string;
  private readonly fetchImpl: typeof fetch;

  constructor(config: { baseUrl: string; bearer: string; fetch?: typeof fetch }) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.bearer = config.bearer;
    this.fetchImpl = config.fetch ?? fetch;
  }

  async *stream(runId: string, sessionId: string | null, opts: SseOptions = {}): AsyncIterable<SseEvent> {
    const url = new URL(`${this.baseUrl}${RUN_EVENTS_PATH(runId)}`);
    if (opts.since) url.searchParams.set("since", opts.since);

    const headers: Record<string, string> = {
      "Accept": "text/event-stream",
      "Authorization": `Bearer ${this.bearer}`,
      [HEADER_PROTOCOL_VERSION]: PROTOCOL_VERSION,
    };
    if (sessionId) headers[HEADER_SESSION_ID] = sessionId;

    const resp = await this.fetchImpl(url.toString(), {
      method: "GET",
      headers,
      ...(opts.signal ? { signal: opts.signal } : {}),
    });

    if (!resp.ok) {
      throw new AEPTransportError(`SSE HTTP ${resp.status} ${resp.statusText}`);
    }
    if (!resp.body) {
      throw new AEPTransportError("SSE response has no body");
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder("utf-8");
    let buf = "";

    try {
      while (true) {
        const { value, done } = await reader.read();
        if (done) break;
        buf += decoder.decode(value, { stream: true });

        let idx: number;
        while ((idx = buf.indexOf("\n\n")) !== -1) {
          const frame = buf.slice(0, idx);
          buf = buf.slice(idx + 2);
          const parsed = parseFrame(frame);
          if (parsed) yield parsed;
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}

function parseFrame(raw: string): SseEvent | null {
  let eventType = "message";
  let dataLine = "";
  for (const line of raw.split("\n")) {
    if (line.startsWith("event:")) eventType = line.slice(6).trim();
    else if (line.startsWith("data:")) dataLine += line.slice(5).trim();
  }
  if (!dataLine) return null;
  try {
    const parsed = JSON.parse(dataLine);
    return {
      event_id: String(parsed.event_id ?? ""),
      run_id: String(parsed.run_id ?? ""),
      type: String(parsed.type ?? eventType),
      emitted_at: String(parsed.emitted_at ?? ""),
      payload: parsed.payload ?? parsed,
    };
  } catch {
    return null;
  }
}
